File: my-folder/0210-course-schedule-ii/solution.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<Integer>[] graph = new List[numCourses];
        for(int i=0;i<numCourses;i++){
            graph[i] = new ArrayList<>();
        }
        int[] indegree = new int[numCourses];
        for(int[] pair: prerequisites){
            // u->v
            int u = pair[1];
            int v = pair[0];
            graph[u].add(v);
            indegree[v]++;
        }
        Queue<Integer> q = new LinkedList<>();
        for(int i=0;i<numCourses;++i){
            if(indegree[i]==0){
                q.add(i);
            }
        }
        int[] sorted = new int[numCourses];
        int i=0;
        while(!q.isEmpty()){
            int node = q.poll();
            sorted[i++] = node;
            for(int j: graph[node]){
                indegree[j]--;
                if(indegree[j]==0){
                    q.offer(j);
                }
            }
        }
        if(i==numCourses){
            return sorted;
        }
        return new int[0];
    }
}

File: my-folder/0210-course-schedule-ii/solution.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        self.buildgraph(numCourses, prerequisites)
        self.onPath = [False]*numCourses
        self.hasCycle = False
        self.visited = [False]*numCourses
        self.postorder = []
        for i in range(numCourses):
            self.traverse(numCourses,i)
        if self.hasCycle:
            return []
        self.postorder.reverse()
        return self.postorder
        
    def traverse(self, numCourses, s):
        if self.onPath[s]:
            self.hasCycle = True
            
        if self.hasCycle or self.visited[s]:
            return 
        
        self.onPath[s] = True
        self.visited[s] = True

        
        for t in self.graph[s]:
            self.traverse(numCourses,t)
        
        self.postorder.append(s)
        self.onPath[s] = False
        
    def buildgraph(self, numCourses, prerequisites):
        self.graph = []
        for i in range(numCourses):
            self.graph.append([])
            
        for item in prerequisites:
            s = item[1]
            t = item[0]
            self.graph[s].append(t)
        
        
